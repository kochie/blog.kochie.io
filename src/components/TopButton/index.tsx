'use client'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faArrowToTop } from '@fortawesome/pro-duotone-svg-icons'
import { useCallback, useEffect, useState } from 'react'

export default function TopButton() {
  const [visible, setVisibility] = useState(false)
  const [atTop, setTop] = useState(true)
  const [pc, setPc] = useState(0)

  const scrollListener = useCallback(() => {
    if (window.scrollY > 0 && atTop) setTop(false)
    if (window.scrollY === 0 && !atTop) setTop(true)

    const p = document.body.parentNode
    if (!p) return
    // @ts-expect-error the definitions seem to be wrong
    setPc(p.scrollTop / (p.scrollHeight - p.clientHeight))
  }, [atTop])

  useEffect(() => {
    window.addEventListener('scroll', scrollListener)

    return () => {
      window.removeEventListener('scroll', scrollListener)
    }
  }, [scrollListener])

  useEffect(() => {
    setVisibility(true)
  }, [])

  return (
    <div className={visible ? 'visible' : 'invisible'}>
      <div
        className={`fixed bottom-6 right-6 h-20 w-20 bg-green-500 rounded-full z-10 group ${
          atTop ? 'animate-bounce-out' : 'animate-bounce-in'
        }`}
      >
        <div style={{ transform: 'rotate(90deg) scaleX(-1)' }}>
          <svg viewBox="0 0 50 50">
            <circle
              className="progress-circle"
              cx="25"
              cy="25"
              r="22"
              fill="transparent"
              stroke="red"
              strokeWidth={6}
              strokeDasharray={138.16}
              strokeDashoffset={138.16 * pc}
            />
          </svg>
        </div>
        <div
          className={`fixed top-2 left-2 bg-white rounded-full h-16 w-16 shadow-2xl cursor-pointer`}
          onClick={() => {
            window.scrollTo({ top: 0, behavior: 'smooth' })
          }}
        >
          <FontAwesomeIcon
            icon={faArrowToTop}
            className="fa-stack-1x group-hover:animate-bounce-orig text-black"
            size="2x"
          />
        </div>
      </div>
    </div>
  )
}
