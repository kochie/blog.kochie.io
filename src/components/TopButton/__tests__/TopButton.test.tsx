import { ReactTestRenderer, act, create } from 'react-test-renderer'

import TopButton from '..'

describe('Social Media Button Component', () => {
  test('renders correctly', () => {
    let tree: ReactTestRenderer

    act(() => {
      tree = create(<TopButton />)
    })

    // @ts-expect-error tree will be assigned
    expect(tree.toJSON()).toMatchSnapshot()
  })
})
