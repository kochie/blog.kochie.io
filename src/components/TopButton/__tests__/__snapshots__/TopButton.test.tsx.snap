// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Social Media Button Component renders correctly 1`] = `
<div
  className="visible"
>
  <div
    className="fixed bottom-6 right-6 h-20 w-20 bg-green-500 rounded-full z-10 group animate-bounce-out"
  >
    <div
      style={
        {
          "transform": "rotate(90deg) scaleX(-1)",
        }
      }
    >
      <svg
        viewBox="0 0 50 50"
      >
        <circle
          className="progress-circle"
          cx="25"
          cy="25"
          fill="transparent"
          r="22"
          stroke="red"
          strokeDasharray={138.16}
          strokeDashoffset={0}
          strokeWidth={6}
        />
      </svg>
    </div>
    <div
      className="fixed top-2 left-2 bg-white rounded-full h-16 w-16 shadow-2xl cursor-pointer"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-arrow-up-to-line fa-2x fa-stack-1x group-hover:animate-bounce-orig text-black"
        data-icon="arrow-up-to-line"
        data-prefix="fad"
        focusable="false"
        role="img"
        style={{}}
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          className="fa-duotone-group"
          style={{}}
        >
          <path
            className="fa-secondary"
            d="M0 64C0 46.3 14.3 32 32 32H352c17.7 0 32 14.3 32 32s-14.3 32-32 32H32C14.3 96 0 81.7 0 64z"
            fill="currentColor"
            style={{}}
          />
          <path
            className="fa-primary"
            d="M169.4 169.4c12.5-12.5 32.8-12.5 45.3 0l128 128c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L224 269.3V448c0 17.7-14.3 32-32 32s-32-14.3-32-32V269.3L86.6 342.6c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3l128-128z"
            fill="currentColor"
            style={{}}
          />
        </g>
      </svg>
    </div>
  </div>
</div>
`;
